### PR TITLE
504 bloulation selector

### LINKED ISSUE #504: VMD blobulator needs a residue selection
Proteins with massive chains can take up to 30 seconds to process one command. Adding a window to blobulate can reduce processing time while also blobulating only relevant sections.

### LINKED ISSUE #509: VMD blobulator needs new threshold names and GUI redesign
need to move sections of the GUI around

### LINKED ISSUE #512: VMD: scale inversion?
Something odd happens when I change hydrophobicity scales.
With the KD scale, (8D66) looks normal:
image
But the Eisenberg-Weiss and Moon-Fleming scales appear inverted:
image
image

@RL0109 can you confirm?

### LINKED ISSUE #517: Blobulation doesn't update if textbox is changed directly

### LINKED ISSUE #518: Hydropathy default will not change unless initially blobulated

### LINKED ISSUE #523: BlobGUI: Moon-Fleming and Eisenberg-Weiss scales are inverted
image
image

### LINKED ISSUE #538: Blobulator plugin has incorrect sig figs

### PR DESCRIPTION
**Pull request by RL0109**
This is a branch off namespaceGUI not main
Closes #504 - branch has created a working selector function
Closes #509 - Threshold names have been changed to length and hydrophobicity
Closes #512 - Scale inversion has been fixed
Closes #517 - You can now press enter in the textbox to blobulate
Closes #518 -  Default works, even if blobulate hasn't been pressed
Closes #523 - Scale inversion has been fixed
Closes #538 - H default values has proper sig figs

### CODE REVIEW

File: blobulator/compute_blobs.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 from .amino_acids import (
     properties_charge,
     THREE_TO_ONE,
-    properties_type,
+    properties_type,
```
Comment (EzryStIago) [inline blobulator/compute_blobs.py:6]: @RL0109 changes to this file need to be reverted

File: blobulator/compute_blobs.py [hunk 2]
```diff
@@ -516,7 +516,7 @@ def clean_df(df):
     #df['Kyte-Doolittle_hydropathy'] = df['Normalized_Kyte-Doolittle_hydropathy']*9-4.5
 
     return df
-
+protein_1 = "AAAAAAAMMMMMMM"
```
Comment (EzryStIago) [inline blobulator/compute_blobs.py:519]: @RL0109 changes to this file need to be reverted

File: VMD_scripts/Blob_GUI.tcl [hunk 3]
```diff
@@ -5,21 +5,45 @@ if [winfo exists .blob] {
...
 	variable blobColorType1 "Blob Type"
 	variable blobColorType2 "Blob ID"
 	variable hydropathyScale1 "Kyte-Doolittle"
 	variable hydropathyScale2 "Eisenberg-Weiss"
 	variable hydropathyScale3 "Moon-Fleming"
 	variable graphRepOptions "Blob ID"
+
+
+	#Switch variables and default values 
+	variable hydropathyScaleDictionaryList "Kyte-Doolittle"
```
Comment (EzryStIago) [inline VMD_scripts/Blob_GUI.tcl:37]: What is this variable for?

File: VMD_scripts/Blob_GUI.tcl [hunk 4]
```diff
@@ -5,21 +5,45 @@ if [winfo exists .blob] {
...
+	variable resStart 
+	variable resEnd
 	variable Lmin 4
 	variable H .4
 	variable MolID 
+	variable isFirst 0
```
Comment (EzryStIago) [inline VMD_scripts/Blob_GUI.tcl:43]: What's this?

File: VMD_scripts/Blob_GUI.tcl [hunk 5]
```diff
@@ -32,30 +56,62 @@ variable blobs [toplevel ".blob"]
...
+	-variable ::blobulator::checkForSelect ]  -row 0 -column $::blobulator::checkBoxColumn -columnspan 2 
+
+	#grid [columnconfigure $blobulator::blobs $::blobulator::checkBoxColumn -uniform]
+
+	#Hydropathy Scale grids
+	grid [label $::blobulator::blobs.t2 -text "         Hydropathy Scale:                                                              " -height 2] -row 2 -column $::blobulator::text2Column -columnspan 2
+	grid [checkbutton $::blobulator::blobs.check -text "Auto Updates Hydrophobicity         " \
+	 -variable ::blobulator::checkForUpdate -command {::blobulator::blobulationSlider }] -row 2 -column $::blobulator::checkBoxColumn -columnspan 2 -sticky w
+	grid [ttk::combobox $::blobulator::blobs.dmnu2  -textvariable ::blobulator::hydropathyScaleDictionaryList -width $::blobulator::dropDownMenuWidth \
+	 -values [list $::blobulator::hydropathyScale1 $::blobulator::hydropathyScale2 $::blobulator::hydropathyScale3] -state readonly] -pady 6 -row 2 -column $::blobulator::dropDownColumn -columnspan 2 -sticky w
```
Comment (EzryStIago) [inline VMD_scripts/Blob_GUI.tcl:78]: scale1? scale2? I don't understand this line

File: VMD_scripts/Blob_GUI.tcl [hunk 6]
```diff
@@ -32,30 +56,62 @@ variable blobs [toplevel ".blob"]
...
 		set ::blobulator::MolID "top"
 	}
-	set paraList [list Lmin ::blobulator::Lmin 1 50 1 H ::blobulator::H .1 1 .01]
-	foreach { entry variableName min max interval} $paraList {
-		set w1 [label $::blobulator::blobs.l_$entry -text $entry]
-		set w2 [entry $::blobulator::blobs.e_$entry -width 10 -textvariable $variableName]
-		set w3 [scale $::blobulator::blobs.s_$entry -orient horizontal -from $min -to $max -length 175 -resolution $interval -tickinterval 0 -variable $variableName -showvalue 0]
+
+	#Atomselect grids
+	grid [label $::blobulator::blobs.lselect -text "         Selection:" ] -row 0 -column $::blobulator::text2Column -sticky e
```
Comment (EzryStIago) [inline VMD_scripts/Blob_GUI.tcl:66]: You're addressing a lot of disparate issues in this PR. Which issue are these changes addressing?

File: VMD_scripts/Blob_GUI.tcl [hunk 7]
```diff
@@ -73,13 +129,23 @@ proc ::blobulator::GUI {} {
 #	::blobulator::blobColorType2 (String): A parameter used for ::blobulator::graphRepOptions checks, if it is set to this variable, will call ::blobulator::graphRepUser2 proc
 #	blobs (Object): Overarching window frame
 proc blobulation {} {
+	if {$::blobulator::checkForSelect == 1} {
```
Comment (EzryStIago) [inline VMD_scripts/Blob_GUI.tcl:132]: would you ever not check for a selection?

File: VMD_scripts/Blob_GUI.tcl [hunk 8]
```diff
@@ -111,7 +184,25 @@ return
 #	::blobulator::blobColorType2 (String): A parameter used for ::blobulator::graphRepOptions checks, if it is set to this variable, will call ::blobulator::graphRepUser2 proc
 proc ::blobulator::blobulationSlider {} {
 	if {$::blobulator::isFirst == 1} {
```
Comment (EzryStIago) [inline VMD_scripts/Blob_GUI.tcl:186]: is first what?

File: VMD_scripts/Blob_GUI.tcl [hunk 9]
```diff
@@ -214,11 +305,13 @@ proc ::blobulator::graphRepUser {} {
...
 	set sel [atomselect $::blobulator::MolID protein]
+	puts [$sel get user2]
 	set user2length [lsort -unique [$sel get user2]]
-	$sel delete
+	puts $user2length
+	$sel delete 
 	foreach u2 $user2length {
 
-		mol representation QuickSurf 1.21 1 .5 2
+		mol representation QuickSurf 1.21 1 .5 3
```
Comment (EzryStIago) [inline VMD_scripts/Blob_GUI.tcl:314]: what's the purpose of this change?

File: VMD_scripts/blobulation.tcl [hunk 10]
```diff
@@ -89,23 +89,25 @@ proc ::blobulator::blobulate {MolID lMin H dictInput} {
...
-	set hsblob [ ::blobulator::hsBlob  $hblob $digitized $lMin ]
-	set hpsblob [ ::blobulator::hpsBlob  $hsblob $digitized ]
-	set groupedBlob [::blobulator::blobGroup $hpsblob]
-	set blobulated [::blobulator::blobAssign $hpsblob]
-	set blobIndexList [ ::blobulator::blobIndex $blobulated ]
-	if {$blobulated != -1} {
-		::blobulator::blobUserAssign $blobulated $MolID
-		::blobulator::blobUser2Assign $blobIndexList $MolID
-		blobUser3Assign $groupedBlob $MolID
+
```
Comment (EzryStIago) [inline VMD_scripts/blobulation.tcl:92]: What's the purpose of these changes?

File: VMD_scripts/blobulation.tcl [hunk 11]
```diff
@@ -532,7 +643,7 @@ proc ::blobulator::blobIndex { blob } {
 	
 	
 	set blobChar q
-	set count 1
+	set count 0
```
Comment (EzryStIago) [inline VMD_scripts/blobulation.tcl:646]: why this change?

File: VMD_scripts/blobulation.tcl [hunk 12]
```diff
@@ -620,13 +731,68 @@ proc ::blobulator::blobUserAssign { blob1 MolID } {
 	
  }
 
-proc ::blobulator::blobUser2Assign { blob2 MolID } {
+proc ::blobulator::blobUserAssignSelector {blob1 MolID chainList} {
```
Comment (EzryStIago) [inline VMD_scripts/blobulation.tcl:734]: This function appears to lack documentation

File: VMD_scripts/blobulation.tcl [hunk 13]
```diff
@@ -620,13 +731,68 @@ proc ::blobulator::blobUserAssign { blob1 MolID } {
...
+				set sel2 [atomselect $molid "residue $rs and protein"]
+				$sel2 set user $i
+			}
+			$sel2 delete
+		}
+	}
+		
+}
+
+proc ::blobulator::blobUser2AssignSelector { blob2 MolID chainList} {
```
Comment (EzryStIago) [inline VMD_scripts/blobulation.tcl:763]: This function lacks documentation

**Review by EzryStIago — CHANGES_REQUESTED** [review]
Missing documentation:
1. The PR doesn't have enough information to review. What does the code do now that it didn't do before?
2. There are a lot changes in this PR. I've requested clarifications in several places. Please make smaller PRs in the future - they're easier to review. If this seems cumbersome, let's talk about branching strategies.
3. Some new functions lack documentation headers.

File: VMD_scripts/blobulation.tcl (hunk 10 above)
Comment (EzryStIago) [reply VMD_scripts/blobulation.tcl:92]: referring to the block of changes that start on this line

File: VMD_scripts/Blob_GUI.tcl (hunk 3 above)
Comment (RL0109) [reply VMD_scripts/Blob_GUI.tcl:37]: For GUI's it helps to have default values that are instantly selected, less initial work for the user. This sets the hydrophobicity scale to Kyte-Doolittle on startup.

File: VMD_scripts/Blob_GUI.tcl (hunk 9 above)
Comment (RL0109) [reply VMD_scripts/Blob_GUI.tcl:314]: This increases the resolution of the quicksurf represnetations.

File: VMD_scripts/Blob_GUI.tcl (hunk 4 above)
Comment (RL0109) [reply VMD_scripts/Blob_GUI.tcl]: Changed variable to isFirstTimeBlobulating

File: VMD_scripts/Blob_GUI.tcl (hunk 5 above)
Comment (RL0109) [reply VMD_scripts/Blob_GUI.tcl:78]: These scales are string values with the hydrophobicity scale names. This generates a list that the drop down menu will use.

File: VMD_scripts/Blob_GUI.tcl (hunk 6 above)
Comment (RL0109) [reply VMD_scripts/Blob_GUI.tcl:66]: This issue was adressing UI changes. The spaces were to push the text further in the grid to align with parallel text.

File: VMD_scripts/blobulation.tcl (hunk 13 above)
Comment (RL0109) [reply VMD_scripts/blobulation.tcl:783]: Documentation has been added

File: VMD_scripts/blobulation.tcl (hunk 12 above)
Comment (RL0109) [reply VMD_scripts/blobulation.tcl:747]: Documentation has been added

File: VMD_scripts/blobulation.tcl (hunk 11 above)
Comment (RL0109) [reply VMD_scripts/blobulation.tcl:646]: This was causing an issue with blob ID's. Off by 1 error.

File: VMD_scripts/Blob_GUI.tcl (hunk 7 above)
Comment (RL0109) [reply VMD_scripts/Blob_GUI.tcl:132]: Yes, this variable is checking if the checkbox for the selector is active or not.

File: VMD_scripts/blobulation.tcl (hunk 10 above)
Comment (RL0109) [reply VMD_scripts/blobulation.tcl:92]: I believe no changes actually happened, I think I cut by accident and pasted directly after.

**Comment by cpitman-11** [comment]
When h-blobs detected within the atomselect extend outside of the atomselect, they still appear on the structure. Is this expected?
Screenshot 2024-11-01 at 2 13 51 PM

**Comment by cpitman-11** [comment]
An error gets thrown if the user deletes the text in the hydrophobicity threshold/length minimum cutoff boxes to change the value

**Comment by cpitman-11** [comment]
Other than these two things it seems to be working well/as expected

File: VMD_scripts/Blob_GUI.tcl (hunk 5 above)
Comment (EzryStIago) [reply VMD_scripts/Blob_GUI.tcl:81]: Made a new issue for organization: #551

File: VMD_scripts/blobulation.tcl (hunk 10 above)
Comment (EzryStIago) [reply VMD_scripts/blobulation.tcl:92]: AH it was a whitespace change - you indented.

File: VMD_scripts/blobulation.tcl (hunk 11 above)
Comment (EzryStIago) [reply VMD_scripts/blobulation.tcl:646]: Ah, one too many.

**Review by EzryStIago — APPROVED** [review]
